feat(rxFeedback): rxFeedback Updates & Feedback Documentation

halkon opened the pull request:
Relates to #1162 (it's the evolution of the closed PR)

* Updated layout of feedback
* Updated `rxFeedback` to have a controller
   * Allow controller to call `FeedbackService` if available to let it override `$scope` functionality
* Updated Feedback documentation 

[screen shot 2015-09-22 at 12 45 56 pm](https://cloud.githubusercontent.com/assets/1437154/10037674/fa4feb12-617e-11e5-90d3-40af630a330a.png) *outdated*
[screen shot 2015-09-22 at 12 45 49 pm](https://cloud.githubusercontent.com/assets/1437154/10037676/fa53d3c6-617e-11e5-8b3d-a05b7608bb30.png) *outdated*
[screen shot 2015-09-22 at 12 45 44 pm](https://cloud.githubusercontent.com/assets/1437154/10037675/fa53a1e4-617e-11e5-828c-a3c7dae34423.png) *outdated*
[screen shot 2015-09-22 at 12 45 39 pm](https://cloud.githubusercontent.com/assets/1437154/10037677/fa53f810-617e-11e5-9c21-7cbe2876c127.png) *outdated*

@glynnis this includes some of the changes from Ali, not sure if he's a collaborator here or not, I did not spend a lot of time making them exactly the same but will talk to you in regards to that. 

This PR also depends on: https://github.com/rackerlabs/encore-ui-svcs/pull/162

glynnis commented on src/rxFeedback/templates/feedbackForm.html:
```diff
@@ -1,7 +1,13 @@
+  class="rx-feedback-form"
+  ng-switch="state">
   <rx-form-section>
+    <strong class="title">We want to hear your voice.</strong>
```
The `.title` class is being deprecated. See conversion table here: http://rackerlabs.github.io/encore-ui/#/styles/typography

halkon replied:
keep forgetting about that: 

![image](https://cloud.githubusercontent.com/assets/1437154/10057974/913e51c4-6208-11e5-83e9-b78b14126ba5.png)

glynnis commented on src/rxFeedback/templates/feedbackForm.html:
```diff
@@ -1,7 +1,13 @@
+    <strong class="title">We want to hear your voice.</strong>
     <rx-field>
-      <rx-field-name>Report Type:</rx-field-name>
+      <rx-field-name class="title subdued">Choose a topic:</rx-field-name>
```
The `.title` class is being deprecated. You shouldn't need either `.title` or `.subdued` here -- just used the default styles provided by `<rx-field-name>`.

halkon replied:
that's fine, `subdued` matched the color in the comps and `title` matched more or less the height.

glynnis commented on src/rxFeedback/templates/feedbackForm.html:
```diff
@@ -11,13 +17,18 @@
               ng-init="fields.type = feedbackTypes[0]"
               required="true"></select>
         </rx-input>
+        <rx-help-text ng-if="state === 'redirect'">You'll be redirected to a new window</rx-help-text>
```
:+1:

glynnis replied:
Oh, actually, while I'm being picky: can you add a period to the end of the sentence?

halkon replied:
![fry-not-sure-if](https://cloud.githubusercontent.com/assets/1437154/10058015/c3f02a48-6208-11e5-8c32-0d88b12bf033.jpg)

glynnis commented on src/rxFeedback/templates/feedbackForm.html:
```diff
@@ -11,13 +17,18 @@
-
-  <rx-form-section ng-show="fields.type">
+  <!-- if we need to show the redirect message -->
+  <div class="redirect-description well" ng-show="fields.type" ng-switch-when="redirect">
```
This looks fine visually, but just a note: modal wells have their own class, `.modal-well`. See http://rackerlabs.github.io/encore-ui/#/styles/wells

glynnis replied:
Is `.redirect-description` just present for JS? No styles there, right?

halkon replied:
No styles, it was for testing, I'll remove if you wish

glynnis commented on src/rxFeedback/templates/feedbackForm.html:
```diff
@@ -11,13 +17,18 @@
+  <rx-form-section ng-show="fields.type" ng-switch-default>
     <rx-field>
-      <rx-field-name class="feedback-description">{{fields.type.prompt}}:</rx-field-name>
+      <rx-field-name class="subdued feedback-description">{{fields.type.prompt}}:</rx-field-name>
```
No need for `.subdued` here.

glynnis commented:
@halkon With the latest Framework release, I believe the text areas should no longer be outlined in red on load -- only if they're required and "dirty" with some interaction from the user. Let me know if for some reason that's not the case -- we don't want them to be red on load.

CITguy commented:
@glynnis, you are correct. The fields will no longer default to be outlined in red.

halkon commented:
Rebased with latest master and updated with changes: 

Error border is gone
![image](https://cloud.githubusercontent.com/assets/1437154/10058169/af4d4d4a-6209-11e5-86ba-68da9a595e5b.png)

@glynnis 
This is with `.modal-well`
![image](https://cloud.githubusercontent.com/assets/1437154/10058191/cdeabff8-6209-11e5-9ae1-cccc2e76e47c.png)
This is with `.well`
![image](https://cloud.githubusercontent.com/assets/1437154/10058210/ead36480-6209-11e5-9f5a-73d35d636a12.png)

I left it with `.well` as you said that looks fine, and it looks a bit better to me, neither of them match the padding/margin from the field containers

glynnis commented:
Design LGTM as long as tests are passing :+1:

halkon commented:
Updated per conversation outside.  Wrapping the message in a 

doing:
```html
<rx-form-container ...>
   <div class="modal-well" ...>
...
```

![image](https://cloud.githubusercontent.com/assets/1437154/10058490/5ccf5296-620b-11e5-9da7-81d7d184fcfc.png)

CITguy commented on src/rxFeedback/docs/rxFeedback.midway.js:
```diff
@@ -26,7 +26,8 @@ describe('rxFeedback', function () {
 
-    it('should include the url in the subtitle', function () {
+    // This is being skipped due to removal of the subtitle
+    it.skip('should include the url in the subtitle', function () {
```
If the subtitle is no longer there, why not just remove the test?

CITguy commented on src/rxFeedback/templates/feedbackForm.html:
```diff
@@ -1,7 +1,13 @@
+  class="rx-feedback-form"
+  ng-switch="state">
   <rx-form-section>
+    <h3>We want to hear your voice.</h3>
```
If you wish the header to be flush with the content, you can wrap it in a `<div>`. `rx-form-section rx-field` and `rx-form-section > div` have the [same positioning styles](https://github.com/rackerlabs/encore-ui/blob/master/src/rxForm/rxForm.less#L397-L404).

halkon commented:
@CITguy updated

<img width="570" alt="screen shot 2015-09-28 at 9 47 41 am" src="https://cloud.githubusercontent.com/assets/1437154/10138872/060f2508-65c6-11e5-8203-2542effed85c.png">
<img width="570" alt="screen shot 2015-09-28 at 9 47 23 am" src="https://cloud.githubusercontent.com/assets/1437154/10138871/060e651e-65c6-11e5-8d30-9cc421cf01ac.png">

CITguy commented on src/rxFeedback/README.md:
```diff
@@ -4,11 +4,22 @@ Component built to gather and send user feedback
-This endpoint also supports a `product` parameter `/api/encore/feedback/:product` for sending feedback to a 
-product-specific mailing list. Adding a custom endpoint is managed in `encore-service-pillar`. Once configured 
+The `rxFeedback` component sends feedback to `/api/encore/feedback`, which routes feedback to `encoreui@lists`.
+For the **Feedback Type** of "Feature Request", we will open up a new window redirecting the user to the **GET Feedback** website, which will now host all internal requests for features.
```
Is this the default functionality **without** using `encore.svcs.feedback`?

halkon replied:
nope, I'll update

CITguy commented on src/rxFeedback/README.md:
```diff
@@ -26,3 +37,73 @@ feedback object with the following definition:
+### Production
+To manually include the Feedback changes without updating your version of Encore UI (but after updating `encore-ui-svcs`), please include the following:
+
+Add the following script in your `index.html` (after `encore-ui` and `encore-ui-svcs` **or** after injected dependencies):
```
Is there any functional difference between the two options? I'd rather simplify to `after injected dependencies`.  As it is now, the wording is confusing and the example doesn't reflect the instruction.

```javascript
<!-- endinject -->
<script src="long.cdn.url/feedback-override.js"></script>
```

CITguy commented on src/rxFeedback/README.md:
```diff
@@ -26,3 +37,73 @@ feedback object with the following definition:
+
+-- OR --
+
+Include in your `app.js` is the following snippet:
```
I'd rather provide one option for injection (the CDN option).  The snippet would need updated every time the view changes (no bueno), while the CDN can update with no change to consuming application code.

halkon replied:
good call, removed

CITguy commented on src/rxFeedback/README.md:
````diff
@@ -26,3 +36,62 @@ feedback object with the following definition:
+```javascript
+<!-- inject:js -->
+<!-- endinject -->
+<!-- or -->
````
probably don't need this comment

halkon replied:
Get rid of the `or`? I'm thinking of just having two code blocks.  Thoughts?

CITguy replied:
If we're simplifying to one technique, I'm not sure the `or` comment is necessary. Just need...

```html
<!-- inject:js -->
<!-- endinject -->
<script src="https://6618f7541d71c1a404be-a22eac1892b2a6dcfdb36104c0e925de.ssl.cf1.rackcdn.com/feedback-override.js"></script>
```

Am I understanding this correctly?

halkon replied:
you are, essentially most projects based on the generator will have the injection, previous projects will have direct inclusion of `encore-ui` and `encore-ui-svcs`, that's why I was including both.  

However you're right if we're boiling it down to just one, we should do the latter one and not have mention of manual includes

CITguy commented on src/rxFeedback/README.md:
```diff
@@ -4,11 +4,21 @@ Component built to gather and send user feedback
+
+This endpoint also supports a `product` parameter `/api/encore/feedback/:product` for sending feedback to a
+product-specific mailing list.
+
```
I'd add a `## Custom Endpoint` heading here.

halkon replied:
Added

CITguy commented on src/rxFeedback/README.md:
````diff
@@ -26,3 +36,62 @@ feedback object with the following definition:
+
+
+```
+angular.module('myApplication', ['ngRoute', 'ngResource', 'encore.svcs.feedback'])
````
So, once you inject `encore.svcs.feedback`, how do you use it?

halkon replied:
you don't

The updates to the controller in for `rxFeedback` do a `.has('FeedbackService')` check, once this dependency is added, that becomes available and gets called.

CITguy replied:
I'd document this somewhere. Otherwise, people will be asking the same question.

CITguy commented on src/rxFeedback/templates/feedbackForm.html:
```diff
@@ -11,11 +19,18 @@
-
-  <rx-form-section ng-show="fields.type">
+  <rx-form-section ng-show="fields.type" ng-switch-when="redirect">
+    <!-- if we need to show the redirect message -->
```
comment probably not necessary

CITguy commented on src/rxFeedback/templates/feedbackForm.html:
```diff
@@ -11,11 +19,18 @@
+      Popup Blocker? <a href="{{ route }}" target="_blank">Go to the page directly.</a>
+    </div>
+  </rx-form-section>
+  <!-- if we need to show the redirect message -->
```
comment probably not necessary

CITguy commented on src/rxFeedback/templates/feedbackForm.html:
```diff
@@ -30,3 +45,7 @@
 </rx-modal-form>
+<rx-modal-footer state="redirect">
+  <button class="button submit" ng-click="cancel()">Continue</button>
+  <button class="button cancel" ng-click="cancel()">Cancel</button>
```
Why do we have two buttons that cancel?

halkon replied:
Good question, the behavior is the same, @glynnis should we just remove the cancel button?

halkon replied:
@CITguy we're gonna wait on Ali, the UX representative that did the layout work on this, to come up with a solution to this question, ETA likely tomorrow morning.

halkon replied:
@CITguy after speaking with Ali, we removed the auto open of a window when on the `Feature Request` option, and only going to open it when `Continue` button or the `link` above is clicked.

CITguy commented:
LGTM.

Closed all visual regression PRs as they all did not pertain to rxFeedback (and they were all very different from each other).  This PR is ready to merge once rackerlabs/encore-ui-svcs#162 has been merged and released.
